lib/consensus/raft/consensus.go

```diff
@@ -0,0 +1,563 @@
+// Package raft implements a Consensus component for IPFS Cluster which uses
+// Raft (go-libp2p-raft).
+package consensus
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/google/uuid"
+
+	addr "github.com/filecoin-project/go-address"
+
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/node/config"
+
+	//ds "github.com/ipfs/go-datastore"
+	logging "github.com/ipfs/go-log/v2"
+	consensus "github.com/libp2p/go-libp2p-consensus"
+	rpc "github.com/libp2p/go-libp2p-gorpc"
+	libp2praft "github.com/libp2p/go-libp2p-raft"
+	host "github.com/libp2p/go-libp2p/core/host"
+	peer "github.com/libp2p/go-libp2p/core/peer"
+)
+
+var logger = logging.Logger("raft")
+
+type RaftState struct {
+	NonceMap map[addr.Address]uint64
+	MsgUuids map[uuid.UUID]*types.SignedMessage
+}
+
+func newRaftState() RaftState {
+	return RaftState{NonceMap: make(map[addr.Address]uint64),
+		MsgUuids: make(map[uuid.UUID]*types.SignedMessage)}
+}
+
+type ConsensusOp struct {
+	Nonce     uint64               `codec:"nonce,omitempty"`
+	Uuid      uuid.UUID            `codec:"uuid,omitempty"`
+	Addr      addr.Address         `codec:"addr,omitempty"`
```

Comment:
Don't we have the address/nonce in the SignedMsg?

Reply:
We do. I figured it'd be nice to have separately since those are the core parts of the state/op but I'm ok just leaving it in SignedMessage if you think its a concern

Reply:
I'd drop those fields, as it's just more state that needs to be managed